File: src/main/java/mnm/mods/tabbychat/extra/filters/FilterAddon.java
package mnm.mods.tabbychat.extra.filters;

import com.google.common.base.Joiner;
import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import mnm.mods.tabbychat.ChatManager;
import mnm.mods.tabbychat.TabbyChatClient;
import mnm.mods.tabbychat.api.events.ChatMessageEvent.ChatReceivedEvent;
import mnm.mods.tabbychat.api.filters.Filter;
import mnm.mods.tabbychat.api.filters.FilterEvent;
import mnm.mods.tabbychat.settings.ServerSettings;
import net.minecraft.client.Minecraft;
import net.minecraftforge.eventbus.api.SubscribeEvent;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Supplier;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class FilterAddon {

    private static Map<String, Optional<Supplier<String>>> variables = Maps.newHashMap();

    private List<Filter> filters = Lists.newArrayList();

    public FilterAddon(ChatManager chat) {
        filters.add(new ChannelFilter(chat));
        filters.add(new MessageFilter(chat));

        variables.clear();

        Minecraft mc = Minecraft.getInstance();
        setVariable("player", () -> Pattern.quote(mc.getSession().getUsername()));
        setVariable("onlineplayer", () -> Joiner.on('|')
                .appendTo(new StringBuilder("(?:"), mc.world.playerEntities.stream()
                        .map(player -> Pattern.quote(player.getName().getString()))
                        .iterator())
                .append(')').toString()
        );
    }

    private static void setVariable(String key, Supplier<String> supplier) {
        variables.put(key, Optional.of(supplier));
    }

    static String getVariable(String key) {
        return variables.getOrDefault(key, Optional.empty())
                .map(Supplier::get)
                .orElse("");
    }

    @SubscribeEvent
    public void onChatRecieved(ChatReceivedEvent message) {
        ServerSettings settings = TabbyChatClient.getInstance().getServerSettings();
        if (settings == null) {
            // We're possibly not in game.
            return;
        }

        for (Filter filter : Iterables.concat(filters, settings.filters)) {
            Matcher matcher = filter.getPattern().matcher(filter.prepareText(message.text));
            while (matcher.find()) {
                FilterEvent event = new FilterEvent(matcher, message.channels, message.text);
                filter.action(event);
                message.text = event.text; // Set the new chat
                message.channels = event.channels; // Add new channels.
            }
        }
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/extra/filters/FilterAddon.kt
package mnm.mods.tabbychat.extra.filters

import mnm.mods.tabbychat.MODID
import mnm.mods.tabbychat.api.events.ChatMessageEvent.ChatReceivedEvent
import mnm.mods.tabbychat.api.filters.FilterEvent
import mnm.mods.tabbychat.client.TabbyChatClient
import mnm.mods.tabbychat.util.mc
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.client.event.ClientPlayerNetworkEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod
import java.util.regex.Pattern

@Mod.EventBusSubscriber(modid = MODID, value = [Dist.CLIENT])
object FilterAddon {

    private val builtinFilters = listOf(ChannelFilter(), MessageFilter())

    val variables = mapOf(
            "player" to { Pattern.quote(mc.session.username) },
            "onlineplayer" to {
                mc.connection!!.playerInfoMap.joinToString("|", "(?:", ")") {
                    Pattern.quote(it.gameProfile.name)
                }
            }
    )

    private val filters = mutableMapOf<String, UserFilter>()

    fun createFilter(name: String): UserFilter {
        require(name !in filters) { "Config with name: '$name' already exists" }

        val config = UserFilter(FilterConfig(name))

        filters[name] = config
        TabbyChatClient.serverSettings.filters.add(config.config)

        return config
    }

    private fun filters() = (builtinFilters + filters.values).asIterable()

    @SubscribeEvent
    fun onChatRecieved(message: ChatReceivedEvent) {
        // We're possibly not in game.
        if (mc.connection == null) return

        val text = message.text
        if (text != null) {

            for (filter in filters()) {


                val matcher = filter.expression.find(filter.prepareText(text))
                if (matcher != null) {
                    val event = FilterEvent(matcher, message.channels, text)
                    filter.action(event)
                    message.text = event.text // Set the new chat
                    message.channels = event.channels // Add new channels.
                }
            }
        }
    }

    @SubscribeEvent
    fun onLogIn(event: ClientPlayerNetworkEvent.LoggedInEvent) {
        filters.clear()
        filters += TabbyChatClient.serverSettings.filters.map {
            it.name to UserFilter(it)
        }
    }

    @SubscribeEvent
    fun onLogOut(event: ClientPlayerNetworkEvent.LoggedOutEvent) {
        filters.clear()
    }
}

